source/common/ur_pool_manager.hpp

```diff
@@ -0,0 +1,179 @@
+    ret = urDeviceGetNativeHandle(rhs.hDevice, &rhsNative);
+    if (ret != UR_RESULT_SUCCESS) {
+        throw ret;
+    }
+
+    return lhsNative == rhsNative && lhs.type == rhs.type &&
+           (isSharedAllocationReadOnlyOnDevice(lhs) ==
+            isSharedAllocationReadOnlyOnDevice(rhs)) &&
+           lhs.poolHandle == rhs.poolHandle;
+};
+
+inline std::size_t pool_descriptor::hash(const pool_descriptor &desc) {
+    ur_native_handle_t native;
+    auto ret = urDeviceGetNativeHandle(desc.hDevice, &native);
+    if (ret != UR_RESULT_SUCCESS) {
+        throw ret;
+    }
+
+    return combine_hashes(0, desc.type, native,
+                          isSharedAllocationReadOnlyOnDevice(desc),
+                          desc.poolHandle);
+};
+
+static inline void
+addPoolDescriptorForHost(ur_usm_pool_handle_t poolHandle,
+                         ur_context_handle_t hContext,
+                         std::vector<pool_descriptor> &descriptors) {
+    pool_descriptor &desc = descriptors.emplace_back();
+    desc.poolHandle = poolHandle;
+    desc.hContext = hContext;
+    desc.type = UR_USM_TYPE_HOST;
+}
+
+static inline void addPoolDescriptorsForDevice(
+    ur_usm_pool_handle_t poolHandle, ur_context_handle_t hContext,
+    ur_device_handle_t hDevice, std::vector<pool_descriptor> &descriptors) {
+    {
+        pool_descriptor &desc = descriptors.emplace_back();
+        desc.poolHandle = poolHandle;
+        desc.hContext = hContext;
+        desc.type = UR_USM_TYPE_DEVICE;
+    }
+    {
+        pool_descriptor &desc = descriptors.emplace_back();
+        desc.poolHandle = poolHandle;
+        desc.hContext = hContext;
+        desc.type = UR_USM_TYPE_SHARED;
+        desc.hDevice = hDevice;
+        desc.deviceReadOnly = false;
+    }
+    {
+        pool_descriptor &desc = descriptors.emplace_back();
+        desc.poolHandle = poolHandle;
+        desc.hContext = hContext;
+        desc.type = UR_USM_TYPE_SHARED;
+        desc.hDevice = hDevice;
+        desc.deviceReadOnly = true;
+    }
+}
+
+inline std::vector<pool_descriptor>
+createPoolDescriptors(ur_usm_pool_handle_t poolHandle,
+                      ur_context_handle_t hContext, size_t deviceCount,
+                      const ur_device_handle_t *phDevices) {
+    std::vector<pool_descriptor> descriptors;
+
+    addPoolDescriptorForHost(poolHandle, hContext, descriptors);
+
+    std::function<void(ur_device_handle_t)> addPoolsForDevicesRec =
+        [&](ur_device_handle_t hDevice) {
+            addPoolDescriptorsForDevice(poolHandle, hContext, hDevice,
+                                        descriptors);
+            auto [ret, subDevices] = urGetSubDevices(hDevice);
+            if (ret != UR_RESULT_SUCCESS) {
+                throw ret;
+            }
+            for (auto &subDevice : subDevices) {
+                addPoolsForDevicesRec(subDevice);
+            }
+        };
+
+    for (size_t i = 0; i < deviceCount; i++) {
+        addPoolsForDevicesRec(phDevices[i]);
```

Comment:
try catch?

Reply:
I don't think anything other then bad_alloc can be thrown here. It would be better to handle this in upper layer